tests/integration/v3_failover_test.go

```diff
@@ -0,0 +1,163 @@
+// Copyright 2022 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package integration
+
+import (
+	"bytes"
+	"context"
+	"crypto/tls"
+	"go.etcd.io/etcd/api/v3/v3rpc/rpctypes"
+	clientv3test "go.etcd.io/etcd/tests/v3/integration/clientv3"
+	"testing"
+	"time"
+
+	"go.etcd.io/etcd/client/v3"
+	integration2 "go.etcd.io/etcd/tests/v3/framework/integration"
+	"google.golang.org/grpc"
+)
+
+func TestFailover(t *testing.T) {
+	cases := []struct {
+		name     string
+		testFunc func(*testing.T, *tls.Config, *integration2.Cluster) (*clientv3.Client, error)
+	}{
+		{
+			name:     "create client before the first server down",
+			testFunc: createClientBeforeServerDown,
+		},
+		{
+			name:     "create client after the first server down",
+			testFunc: createClientAfterServerDown,
+		},
+	}
+
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			t.Logf("Starting test [%s]", tc.name)
+			integration2.BeforeTest(t)
+
+			// Launch an etcd cluster with 3 members
+			t.Logf("Launching an etcd cluster with 3 members [%s]", tc.name)
+			clus := integration2.NewCluster(t, &integration2.ClusterConfig{Size: 3, ClientTLS: &integration2.TestTLSInfo})
+			defer clus.Terminate(t)
+
+			cc, err := integration2.TestTLSInfo.ClientConfig()
+			if err != nil {
+				t.Fatal(err)
+			}
+			// Create an etcd client before or after first server down
+			t.Logf("Creating an etcd client [%s]", tc.name)
+			cli, err := tc.testFunc(t, cc, clus)
+			if err != nil {
+				t.Fatalf("Failed to create client: %v", err)
+			}
+			defer cli.Close()
+
+			// Sanity test
+			key, val := "key1", "val1"
+			retryCount := 10
+			checkRetry := func(err error) bool {
```

Comment:
Please consider factoring these methods to the top-level as: 

`putWithRetries()`
`getWithRetries()`

I think it would improve readability of the test. I understand the desire for test to contain all the testing logic in the single place, but I think in this case, the retry logic is harming the readability around basic etcd operations.

Reply:
Yes, it's valid & good point. Let me update it.

Reply:
Done & thanks.